File: project/jimmer-sql/src/test/java/org/babyfish/jimmer/sql/model/BookStoreAvgPriceResolver.java
package org.babyfish.jimmer.sql.model;

import org.babyfish.jimmer.sql.JSqlClient;
import org.babyfish.jimmer.sql.TransientResolver;
import org.babyfish.jimmer.sql.ast.tuple.Tuple2;
import org.jetbrains.annotations.NotNull;

import java.math.BigDecimal;
import java.sql.Connection;
import java.util.*;

public class BookStoreAvgPriceResolver implements TransientResolver<UUID, BigDecimal> {

    private final JSqlClient sqlClient;

    public BookStoreAvgPriceResolver(JSqlClient sqlClient) {
        this.sqlClient = sqlClient;
    }

    @Override
    public Map<UUID, BigDecimal> resolve(Collection<UUID> ids, Connection con) {
        BookTable book = BookTable.$;
        List<Tuple2<UUID, BigDecimal>> tuples =
                sqlClient.createQuery(book)
                        .where(book.store().id().in(ids))
                        .groupBy(book.store().id())
                        .select(
                                book.store().id(),
                                book.price().avg().coalesce(BigDecimal.ZERO)
                        )
                        .execute(con);
        return Tuple2.toMap(tuples);
    }
}

File: project/jimmer-sql-kotlin/src/test/kotlin/org/babyfish/jimmer/sql/kt/model/BookStoreAvgPriceResolver.kt
package org.babyfish.jimmer.sql.kt.model

import org.babyfish.jimmer.sql.kt.KSqlClient
import org.babyfish.jimmer.sql.kt.KTransientResolver
import org.babyfish.jimmer.sql.kt.ast.expression.avg
import org.babyfish.jimmer.sql.kt.ast.expression.coalesce
import org.babyfish.jimmer.sql.kt.ast.expression.valueIn
import java.math.BigDecimal
import java.sql.Connection

class BookStoreAvgPriceResolver(
    private val sqlClient: KSqlClient
): KTransientResolver<Long, BigDecimal> {

    override fun resolve(
        ids: Collection<Long>,
        con: Connection
    ): Map<Long, BigDecimal> =
        sqlClient
            .createQuery(Book::class) {
                where(table.store.id valueIn ids)
                groupBy(table.store.id)
                select(
                    table.store.id,
                    avg(table.price).coalesce(BigDecimal.ZERO)
                )
            }
            .execute(con)
            .associateBy(
                {it._1}
            ) {
                it._2
            }
}